go/vt/wrangler/reparent.go

```diff
@@ -808,37 +837,71 @@ func (wr *Wrangler) chooseNewMaster(
 	tabletMap map[string]*topo.TabletInfo,
 	avoidMasterTabletAlias *topodatapb.TabletAlias,
 	waitReplicasTimeout time.Duration) (*topodatapb.TabletAlias, error) {
-
-	if avoidMasterTabletAlias == nil {
-		return nil, fmt.Errorf("tablet to avoid for reparent is not provided, cannot choose new master")
+	candidates, err := wr.FindMasterCandidates(ctx, tabletMap, waitReplicasTimeout)
+	if err != nil {
+		return nil, err
 	}
+
 	var masterCell string
 	if shardInfo.MasterAlias != nil {
 		masterCell = shardInfo.MasterAlias.Cell
 	}
 
-	maxPosSearch := maxReplPosSearch{
+	// Filter out avoidMasterTabletAlias, if it exists in the candidates list, and also filter out
+	// any candidates that aren't in the master tablets cell.
+	filteredCandidates := make([]*topodatapb.TabletAlias, 0, len(candidates))
+	for i := range candidates {
+		if candidates[i] != avoidMasterTabletAlias && candidates[i].Cell == masterCell {
+			filteredCandidates = append(filteredCandidates, candidates[i])
+		}
+	}
+
+	if len(filteredCandidates) == 0 {
+		return nil, nil
+	}
+	return filteredCandidates[0], nil
+}
+
+// FindMasterCandidates will look at all of the tablets in the supplied tabletMap, and return a list of tabletAliases that
+// are all caught up. This means that all of the returned TabletAliases will have the same replication position.
+func (wr *Wrangler) FindMasterCandidates(
+	ctx context.Context,
+	tabletMap map[string]*topo.TabletInfo,
+	waitReplicasTimeout time.Duration) ([]*topodatapb.TabletAlias, error) {
+	aggregator := &tabletReplInfoAggregator{
 		wrangler:            wr,
 		ctx:                 ctx,
 		waitReplicasTimeout: waitReplicasTimeout,
 		waitGroup:           sync.WaitGroup{},
-		maxPosLock:          sync.Mutex{},
+		statusLock:          sync.Mutex{},
+		tabletReplInfos:     make([]*replicationInfo, 0, len(tabletMap)),
 	}
 	for _, tabletInfo := range tabletMap {
-		if (masterCell != "" && tabletInfo.Alias.Cell != masterCell) ||
-			topoproto.TabletAliasEqual(tabletInfo.Alias, avoidMasterTabletAlias) ||
-			tabletInfo.Tablet.Type != topodatapb.TabletType_REPLICA {
+		if tabletInfo.Tablet.Type != topodatapb.TabletType_REPLICA {
 			continue
 		}
-		maxPosSearch.waitGroup.Add(1)
-		go maxPosSearch.processTablet(tabletInfo.Tablet)
+		aggregator.waitGroup.Add(1)
+		go aggregator.processTablet(tabletInfo.Tablet)
 	}
-	maxPosSearch.waitGroup.Wait()
+	aggregator.waitGroup.Wait()
 
-	if maxPosSearch.maxPosTablet == nil {
-		return nil, nil
+	sort.Sort(aggregator)
+
+	if len(aggregator.tabletReplInfos) < 2 {
+		// Either 1 result, or 0. Regardless, we can return the list at this point as we don't need to
+		// filter out tablets that aren't fully caught up.
```

Comment:
See above comment. It seems that "fully caught up" should be changed to "the most caught up"

Reply:
Ahh, good point, will update.